### ISSUE TITLE
Can --resource-path be specified multiple times?

### ISSUE THREAD

**Issue by dhimmel** [jgm/pandoc#6152]
Here's the [current documentation](https://github.com/jgm/pandoc/blob/a238128a8e3c8d1942bae26d5eb4d42a8ed2fa39/MANUAL.txt#L881-L896) for the `--resource-path` option:

> `--resource-path=`*SEARCHPATH*
>
> List of paths to search for images and other resources. The paths should be separated by `:` on Linux, UNIX, and macOS systems, and by `;` on Windows. If [`--resource-path`](https://pandoc.org/MANUAL.html#option--resource-path) is not specified, the default resource path is the working directory. Note that, if [`--resource-path`](https://pandoc.org/MANUAL.html#option--resource-path) is specified, the working directory must be explicitly listed or it will not be searched. For example: [`--resource-path=.:test`](https://pandoc.org/MANUAL.html#option--resource-path) will search the working directory and the `test` subdirectory, in that order.
>
> [`--resource-path`](https://pandoc.org/MANUAL.html#option--resource-path) only has an effect if (a) the output format embeds images (for example, `docx`, `pdf`, or `html` with [`--self-contained`](https://pandoc.org/MANUAL.html#option--self-contained)) or (b) it is used together with [`--extract-media`](https://pandoc.org/MANUAL.html#option--extract-media).

The documented method to specify multiple resource paths is not ideal:

- it requires a polymorphic delimiter based on platform
- it is inconsistent with other options that can be specified multiple times to aggregate into an array/list such as `--metadata-file`, `--syntax-definition`, `--include-in-header`, `--bibliography`, `--defaults`.

Therefore, it would be nice to also be able to repeat `--resource-path` in order to define multiple resource paths. See:

```shell
echo "test" | pandoc --to=json --resource-path=. --resource-path=test
```

Currently, this does not create any errors or warnings, but I'm not sure whether one `--resource-path` definition is ignored. So either this feature is already implemented and the docs could be updated or its not implemented and might be a good enhancement?

**Comment by jgm** [comment]
No, the feature is not already implemented; one of these is simply ignored.

> it requires a polymorphic delimiter based on platform

True, but this is the standard way of specifying a search path on each platform.

I'm unconvinced about allowing it to be repeated.  I can imagine, for example, wanting to override a default with a more restrictive search path.

**Comment by dhimmel** [comment]
> No, the feature is not already implemented; one of these is simply ignored.

Ah okay. Maybe a warning would be helpful, but your thinking is that users are intentionally specifying `--resource-path` multiple times from the command line?

> this is the standard way of specifying a search path on each platform

Ah okay, noting that the delimiter is available in Python under [`os.pathsep`](https://docs.python.org/3.8/library/os.html).

> I'm unconvinced about allowing it to be repeated. I can imagine, for example, wanting to override a default with a more restrictive search path.

It seems that other options used to set multiple paths aggregate rather than override.

In the docs for Default files, there's the following:

```yaml
# Note that these take files, not their contents:
include-before-body: []
include-after-body: []
include-in-header: []
resource-path: ["."]
```

Here it seems that `resource-path` should behave similarly to `include-in-header`. Does specifying `--resource-path` overwrite `resource-path` from defaults or append to it?

**Comment by jgm** [comment]
> Does specifying --resource-path overwrite resource-path from defaults or append to it?

Overwrites.

**Comment by jaybe-jekyll** [comment]
`resource-path: [".:my_resources/images:morestuff"` [^1]

Provides for searching within:
- `./`
- `./my_resources/images`
- `./morestuff`

_[^1]: The colon separates entries._

Whereas `include-in-header` can be, e.g.:

```yaml
include-in-header:
- default-headers/init.html
- my-headers/funky.html
```

**Comment by dhimmel** [comment]
> I'm unconvinced about allowing `--resource-path` to be repeated. I can imagine, for example, wanting to override a default with a more restrictive search path.

Overriding can still work properly when multiple `--resource-path` options are aggregated. When a file is searched for in resource-path(s), the file is searched for iteratively and returned upon the first match. Therefore if multiple `--resource-path` directories are supplied there should be no issue as long as they are searched in the correct order.

Imagine the following directory structure:

```tree
├── path-1
│   ├── file-1
│   └── file-2
└── path-2
    └── file-1
```

And the command `pandoc --resource-path=path-1 --resource-path=path-2`.

With the current implementation that `--resource-path=path-1` is ignored:

1. looking for `file-1` will return `path-2/file-1`. This is the overriding case
2. looking for `file-2` will fail

With a potential aggregating implementation:

1. looking for `file-1` will return `path-2/file-1`. The overriding case still works.
2. looking for `file-2` will return `path-1/file-2`

So in both implementations, overriding is possible. If situation 2 arises where `file-2` is searched, I think there's a much large chance the user actually wants `path-1/file-2` than nothing.

In conclusion, I think the iterative search method applied by `--resource-path` means there is little downside, but much to be gained by aggregating rather than overwriting. And there's an opportunity to improve consistency of the API.

**Comment by jaybe-jekyll** [comment]
The described behavior would occur for:

`—resource-path=“[.:assets:sub/assets]”

...and in that order per lookup. 

1. ./
2. ./assets
3. ./sub/assets

Correct?

Adding an additional  —resource-path is then in question as to its placement and prioritization of search paths. 

If this is correct then perhaps an ordering schema:

Adding an *additional * - `—resource-path=“[:another/path]` would *prepend* the added path (note the leading colon). 

Adding another without leading colon appends ... or some variant thereof.

**Comment by jgm** [comment]
@dhimmel I'm talking about overriding a search path with a more restrictive search path, not overriding a file with another file elsewhere in the search path.  E.g. maybe your defaults file says the resource path includes "images", but you want to override that so that "images" is not searched.

**Comment by brainchild0** [comment]
The main question appears to be how to represent a list of strings, specifically path names.

To me the answer is not controversial:

- In the context of environment variables, use the system-dependent path separator (e.g. `:` on Unix).

- In the context of command-line arguments, use repeat specifications of the same argument name.

- In the context of a YAML file, use a list.

Importing one convention into another context seems unnecessary and problematic. 

It is not obvious why it would be useful to override a long list given earlier with a more narrow one. If it is useful, then the mechanics need to be discussed. A solution to whatever specifically is the problem should be possible while keeping the semantics suitable to the context.

**Comment by jgm** [comment]
I'm open to making the change.  I wonder, though, if anyone is relying on the present, overriding behavior.

**Comment by brainchild0** [comment]
Obviously there is no way to know. You could transition through an intermediary phase of warning of future deprecation.

But perhaps the effect is not great, as the usage described strikes me as hard to apply.

From my point of view a user has some number of these options sets. Each one refers to various files and some resource locations. Using a resource location in one set to resolve paths in that same set is much more sensible than mixing the effects so that one set of paths can only be resolved using locations given from a different source.

**Comment by dhimmel** [comment]
> I'm talking about overriding a search path with a more restrictive search path, not overriding a file with another file elsewhere in the search path. E.g. maybe your defaults file says the resource path includes "images", but you want to override that so that "images" is not searched.

I don't see a use case for the "more restrictive search path". Search order is what matters.

Take my above example at https://github.com/jgm/pandoc/issues/6152#issuecomment-594907322.  If a document refers to `file-2`, under the current implementation `file-2` will not be found and will cause an error. I don't see why a user would want a file not to be found.

But if specifying `--resource-path`s also appends rather than overwrites, users will get the first path that matches. Doesn't it make sense to fallback to other resource-paths when the top resource-path is missing a file?

Regarding an `image` directory, why would a user want to remove it from the resource paths? If they want to override an image, they can add an additional resource path with the image files to use. But if they haven't including images with higher-precedent in their resource paths, do they really want no image to be returned at all?

**Comment by brainchild0** [comment]
>  Doesn't it make sense to fallback to other resource-paths when the top resource-path is missing a file?

Yes, but I am beginning to think that this answer may best  apply only  to the case when the resource path and the reference to the file are passed from the same options set, that is, within the same defaults file or both being on the command line. Defaults files can derive value from being self-contained units of reproducible effect on the document processing. Mixing effects across container may seem convenient to some for certain cases, but also might   lead to difficulty.

> 
> Regarding an `image` directory, why would a user want to remove it from the resource paths?



The case of images is confusing to me. I understand that some users may get convenience from Pandoc resolving Pandoc files based on Pandoc options, but should not the resolution of resources  within the source document be determined by the source format? Other processing tools and display environments are not using Pandoc options, but are processing the same document formats. It has seemed  to me best if Pandoc has the same behavior as those tools, as determined by the format.

**Comment by brainchild0** [comment]
> Doesn't it make sense to fallback to other resource-paths when the top resource-path is missing a file?

Yes,  I think so.  

Fancier cases, if required, could be handled with specific instructions, for example to clear the list, or for a particular item to be prepended instead of appended.

Ultimately it is hard for me to tease out the cases because of confusion over the intent of this option, as compared to resolving image locations in whatever format-specific  means would be employed by another application reading the same format.

**Comment by jgm** [comment]
Should specifying `--resource-path` multiple times append to the front or the back of the path?  That is, if you do `pandoc --resource-path=a --resource-path=b`, should the path be `a:b` or `b:a`?

**Comment by dhimmel** [comment]
> Should specifying --resource-path multiple times append to the front or the back of the path? 

I think we want the most consistency with other options that can be specified multiple times. From https://pandoc.org/MANUAL.html:

> **--metadata-file**
This option can be used repeatedly to include multiple metadata files; values in files specified later on the command line will be preferred over those specified in earlier files.

> **--defaults**
> When multiple defaults are used, their contents will be combined.
> Settings from the defaults file may be overridden or extended by subsequent options on the command line.

That makes me think `pandoc --resource-path=a --resource-path=b` should search `b` first, then `a`. The complication is that compound file paths have the opposite order, where the paths are searched in order rather than reverse order.

What if `--resource-path=a:b, --resource-path=c:d` gets parsed as `--resource-path=b --resource-path=a --resource-path=d --resource-path=c`, so the order of precedent would be c > d > a > b? You could phase out the current compound paths to minimize this confusion in the future.

**Comment by brainchild0** [comment]
Agree with @dhimmel:

- The combination of sequencing questions and platform dependence makes the separators a good candidate for phasing out (e.g. via printing a deprecation warning).

- Precedence for items earlier in sequence is the common and intuitive pattern, but consistency across the application takes priority. (Is the list in the defaults file following the same order of precedence?)

**Comment by jgm** [comment]
Another thing we should consider:  a user has just asked whether there's a way to include directories that contain colons in the resource path:
<https://groups.google.com/d/msgid/pandoc-discuss/26a48294-b200-4d8f-ad97-2fc781aeaff3%40googlegroups.com>.  If we switched to requiring multiple invocations of `--resource-path` instead of a colon-separated path, that would solve the problem. Unfortunately, it would also be a good idea to respect backwards-compatibility by treating the colon as a separator, even if we make this change.  For now I suppose they can use a defaults file.

**Comment by brainchild0** [comment]
Back compat is critical, but at least serous breakage from CLI changes is limited to scripts. 

In light of the issue, it may seem helpful to plan for a future change. If so, then tossing in a deprecation warning earlier, as early as possible, would mitigate the later difficulty.

For users having problems, though it may seem unappealing, a workaround might be invoking the new behavior based on an undocumented environment variable.

**Comment by jgm** [comment]
One backwards-compatible solution to the problem would be:

* if the argument to `--resource-path` begins with a `:` or `;` (whichever is the relevant path separator for the OS), then append to existing resource path (including default, if this is the first time the option has been specified)
* if it ends with a path separator, then prepend to the existing resource path
* otherwise, replace the existing resource path

**Comment by jgm** [comment]
Thus,
```
--resource-path ":foo" --resource-path "bar:baz:"
```
would give you `bar:baz:.:foo`. And
```
--resource-path ":foo" --resource-path "bar:baz"
```
would just give you `bar:baz`.

**Comment by jgm** [comment]
Though I suppose the problem with this is that it doesn't generalize well to defaults files, where a YAML array is used.

**Comment by brainchild0** [comment]
Largely for reasons discussed, I suggest that the reliance on separators is not an optimal solution for a forward-looking design. Separation of handling by three separate conditions, append, prepend, and replace, strikes me as a very hairy approach that may easily present further difficulties as a design evolves.

Presently, what is the case for supporting all three conditions, and doing so through use of the separator character?

**Comment by jgm** [comment]
> Separation of handling by three separate conditions, append, prepend, and replace

But note that this is something we've always been able to do with PATH in the shell, and it has served well for many decades.  Sometimes you need to ensure that a directory is first in the path, sometimes you want it to be last.  Sometimes you need to fully control the path and want to replace.

```
# replace
$ PATH=foo
# append
$ PATH=$PATH:foo
# prepend
$ PATH=foo:$PATH
```

**Comment by brainchild0** [comment]
> this is something we've always been able to do with PATH

Yes, surely,  but the  appropriateness of following the same pattern in the solution depends on the closeness of the  use cases of resource path compared to executable search path. 


The latter is normally used to find an executable within system or user paths. The context of this operation is that system administrators and users will in general install multiple versions of the same application on the same system, and do so in directories named using basic characters, without white space or punctuation. Environmental variables are generally set  by shell scripts, which execute as sequences of imperative statements.


 It is important to consider how closely the original background contexts are congruent, before imposing the solution of one onto the problem of the other.

**Comment by jgm** [comment]
In the end I went with the simple option of just letting the resource path accumulate, with later invocations prepended to earlier ones.
